libs/geo/src/main/java/org/opensearch/geometry/utils/WellKnownText.java

```diff
@@ -294,23 +305,67 @@ private Geometry parseGeometry(StreamTokenizer stream) throws IOException, Parse
             case "bbox":
                 return parseBBox(stream);
             case "geometrycollection":
-                return parseGeometryCollection(stream);
+                throw new IllegalStateException("Unexpected type: geometrycollection");
             case "circle": // Not part of the standard, but we need it for internal serialization
                 return parseCircle(stream);
         }
         throw new IllegalArgumentException("Unknown geometry type: " + type);
     }
 
+    /**
+     * Parse geometry collection iteratively
+     *
+     * Parsing geometry collection recursively can lead to StackOverflowError when there is a deeply nested structure of GeometryCollection
+     */
     private GeometryCollection<Geometry> parseGeometryCollection(StreamTokenizer stream) throws IOException, ParseException {
         if (nextEmptyOrOpen(stream).equals(EMPTY)) {
             return GeometryCollection.EMPTY;
         }
-        List<Geometry> shapes = new ArrayList<>();
-        shapes.add(parseGeometry(stream));
-        while (nextCloserOrComma(stream).equals(COMMA)) {
-            shapes.add(parseGeometry(stream));
+
+        List<Geometry> topLevelShapes = new ArrayList<>();
+        Stack<List<Geometry>> stack = new Stack<>();
```

Comment:
Oh -- I remembered that `java.util.Stack` is generally considered "deprecated", because it extends the old Java 1.0 `Vector` class, where every operation is `synchronized`.

You should probably change this to:

```
Deque<List<Geometry>> stack = new ArrayDeque<>();
```

Also, you should probably put an upper-bound on the size of the stack as a final safeguard. For example, I think it's perfectly reasonable to say that `GeometryCollections` cannot be nested more than 1000 deep.